proof_of_concept/registry_api.py

```diff
@@ -0,0 +1,111 @@
+
+
+class PartyRegistration:
+    """A handler for the /parties endpoint."""
+    def __init__(self, registry: Registry, deserializer: Deserializer) -> None:
+        """Create a PartyRegistration handler.
+
+        Args:
+            registry: The registry to send requests to.
+            deserializer: A Deserializer to (de)serialise objects with.
+        """
+        self._registry = registry
+        self._deserializer = deserializer
+
+    def on_post(self, request: Request, response: Response) -> None:
+        """Handle a party registration request.
+
+        Args:
+            request: The submitted request.
+            response: A response object to configure.
+
+        """
+        try:
+            self._registry.register_party(self._deserializer(
+                PartyDescription, request.media))
+            response.status = HTTP_201
+            response.body = 'Created'
+        except ValidationError:
+            response.status = HTTP_400
+            response.body = 'Invalid request'
+        except RuntimeError:
+            response.status = HTTP_409
+            response.body = 'Party already exists'
+
+
+class SiteRegistration:
+    """A handler for the /sites endpoint."""
+    def __init__(self, registry: Registry, deserializer: Deserializer) -> None:
+        """Create a SiteRegistration handler.
+
+        Args:
+            registry: The registry to send requests to.
+            deserializer: A Deserializer to deserialise objects with.
+        """
+        self._registry = registry
+        self._deserializer = deserializer
+
+    def on_post(self, request: Request, response: Response) -> None:
+        """Handle a site registration request.
+
+        Args:
+            request: The submitted request.
+            response: A response object to configure.
+
+        """
+        try:
+            # self._registry.register_site(self._deserializer(
+            #     SiteDescription, request.media))
+            response.status = HTTP_201
+            response.body = 'Created'
+        except ValidationError:
+            response.status = HTTP_400
+            response.body = 'Invalid request'
+        except RuntimeError:
+            response.status = HTTP_409
+            response.body = 'Site already exists'
+
+
+class RegistryApi:
+    """The complete Registry REST API.
+
+    Attributes:
+        app: The WSGI application object.
+
+    """
+    def __init__(self, registry: Registry) -> None:
+        """Create a RegistryApi instance.
+
+        Args:
+            registry: The registry to serve for.
+
+        """
+        self.app = App()
+
+        registry_api_file = Path(__file__).parent / 'registry_api.yaml'
+        with open(registry_api_file, 'r') as f:
+            registry_api_def = yaml.safe_load(f.read())
+
+        deserializer = Deserializer(registry_api_def)
+
+        party_registration = PartyRegistration(registry, deserializer)
```

Comment:
Now we have to manually do the routing and we see that we don't fully make use of the power of swagger + connexion. But to be honest, I really don't mind actually, this makes it much less magical.

Reply:
Yeah. If I had the time, I'd write a code generator that generates this code from an OpenAPI spec. Next life...